### PR TITLE
Added Support for Additional Background run time through configuration

### PR DESCRIPTION
**Pull request by desusai7**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-ios/106)

### CODE REVIEW

File: Sources/Classes/Public/RSConfig.h [hunk 1]
```diff
@@ -21,6 +21,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic) int configRefreshInterval;
 @property (nonatomic) bool trackLifecycleEvents;
 @property (nonatomic) bool recordScreenViews;
+@property (nonatomic) bool enableBackGroundRunTime;
```
Comment (pallabmaiti) [inline Sources/Classes/Public/RSConfig.h:24]: Don't like the variable name. Change it to `enableBackgroundMode`

File: Sources/Classes/Public/RSConfigBuilder.h [hunk 2]
```diff
@@ -29,6 +29,7 @@ NS_ASSUME_NONNULL_BEGIN
 - (instancetype) withConfigRefreshInteval: (int) configRefreshInterval;
 - (instancetype) withTrackLifecycleEvens: (BOOL) trackLifecycleEvents;
 - (instancetype) withRecordScreenViews: (BOOL) recordScreenViews;
+- (instancetype) withBackGroundRunTime:(BOOL)enableBackGroundRunTime;
```
Comment (pallabmaiti) [inline Sources/Classes/Public/RSConfigBuilder.h:32]: Rename the function name to 
`- (instancetype) withEnableBackGroundMode:(BOOL) enableBackgroundMode;`

File: Sources/Classes/Public/RSConstants.h [hunk 3]
```diff
@@ -29,6 +29,8 @@ extern NSString *const RSControlPlaneUrl;
 extern bool const RSTrackLifeCycleEvents;
 // whether we should record screen views automatically
 extern bool const RSRecordScreenViews;
+// whether we should add support for background run time
+extern bool const RSEnableBackGroundRunTime;
```
Comment (pallabmaiti) [inline Sources/Classes/Public/RSConstants.h:33]: Change it here as well.

File: Sources/Classes/RSConfig.m [hunk 4]
```diff
@@ -23,6 +23,7 @@ - (instancetype)init
         _configRefreshInterval = RSConfigRefreshInterval;
         _trackLifecycleEvents = RSTrackLifeCycleEvents;
         _recordScreenViews = RSRecordScreenViews;
+        _enableBackGroundRunTime = RSEnableBackGroundRunTime;
```
Comment (pallabmaiti) [inline Sources/Classes/RSConfig.m:26]: Here.

File: Sources/Classes/RSConfigBuilder.m [hunk 5]
```diff
@@ -133,6 +133,14 @@ - (instancetype) withRecordScreenViews:(BOOL)recordScreenViews {
     return self;
 }
 
+- (instancetype) withBackGroundRunTime:(BOOL)enableBackGroundRunTime {
```
Comment (pallabmaiti) [inline Sources/Classes/RSConfigBuilder.m:136]: Here.

File: Sources/Classes/RSConstants.m [hunk 6]
```diff
@@ -17,5 +17,6 @@ @implementation RSConstants
 NSString *const RSControlPlaneUrl = @"https://api.rudderlabs.com";
 bool const RSTrackLifeCycleEvents = YES;
 bool const RSRecordScreenViews = NO;
-NSString *const RS_VERSION = @"1.2.2";
+bool const RSEnableBackGroundRunTime = NO;
```
Comment (pallabmaiti) [inline Sources/Classes/RSConstants.m:20]: Here.

File: Sources/Classes/RSEventRepository.m [hunk 7]
```diff
@@ -49,6 +49,12 @@ - (instancetype)init : (NSString*) _writeKey config:(RSConfig*) _config {
         
         writeKey = _writeKey;
         config = _config;
+        if(config.enableBackGroundRunTime)
```
Comment (pallabmaiti) [inline Sources/Classes/RSEventRepository.m:52]: Here.

File: Sources/Classes/RSEventRepository.m [hunk 8]
```diff
@@ -49,6 +49,12 @@ - (instancetype)init : (NSString*) _writeKey config:(RSConfig*) _config {
...
+        {
+            [RSLogger logDebug:@"EventRepository: Enabling Background run time"];
```
Comment (pallabmaiti) [inline Sources/Classes/RSEventRepository.m:54]: Here.

File: Sources/Classes/RSEventRepository.m [hunk 9]
```diff
@@ -541,12 +547,17 @@ - (void)_applicationWillEnterForeground {
         return;
     }
     
+    if(config.enableBackGroundRunTime) {
```
Comment (pallabmaiti) [inline Sources/Classes/RSEventRepository.m:550]: Here.

File: Sources/Classes/RSEventRepository.m [hunk 10]
```diff
@@ -560,5 +571,24 @@ - (void) __prepareScreenRecorder {
     [UIViewController rudder_swizzleView];
 }
 
+- (void) registerBackGroundTask {
+    [RSLogger logDebug:@"EventRepository: registerBackGroundTask: Registering for Background run time"];
```
Comment (pallabmaiti) [inline Sources/Classes/RSEventRepository.m:575]: Here.

File: Sources/Classes/RSEventRepository.m [hunk 11]
```diff
@@ -560,5 +571,24 @@ - (void) __prepareScreenRecorder {
...
+        [self endBackGroundTask];
+    }];
+}
+
+- (void) endBackGroundTask {
+    [[UIApplication sharedApplication] endBackgroundTask:backgroundTask];
+    backgroundTask = UIBackgroundTaskInvalid;
+}
+
+- (void) reInstateBackGroundTask {
```
Comment (pallabmaiti) [inline Sources/Classes/RSEventRepository.m:586]: Can we merge this function with `registerBackGroundTask`? Make 1 function to `registerBackGroundTask`.
```
if(backgroundTask != UIBackgroundTaskInvalid) {
        [self endBackGroundTask];
    } else {
backgroundTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
        [self endBackGroundTask];
    }];
}
```

**Review by pallabmaiti — CHANGES_REQUESTED** [review]
Can you please do the necessary changes?

**Comment by pallabmaiti** [comment]
Please resolve the conflicts as well?

File: Sources/Classes/RSEventRepository.m (hunk 11 above)
Comment (desusai7) [reply Sources/Classes/RSEventRepository.m]: Yeah, we can do that

**Comment by desusai7** [comment]
@pallabmaiti, addressed the comments, can you please take a look now !

File: Sources/Classes/RSEventRepository.m [hunk 12]
```diff
@@ -595,5 +606,20 @@ - (void) __prepareScreenRecorder {
 #endif
 }
 
+- (void) registerBackGroundTask {
+    [RSLogger logDebug:@"EventRepository: registerBackGroundTask: Registering for Background Mode"];
```
Comment (pallabmaiti) [inline Sources/Classes/RSEventRepository.m:610]: This log will go to line no 614

Comment (desusai7) [reply Sources/Classes/RSEventRepository.m:610]: Oh, Thank you for pointing this out. 
Actually, this function should register a background task every time it is being called, but before doing that it should check if the existing task is still valid and if so invalidate it. So we should pull that logic out of the else block and this log shall be printing always. 
Made the changes and pushed them accordingly.

**Review by pallabmaiti — APPROVED** [review]

**Pull request merged**